### PR TITLE
Loosen absolute tolerance in GBSA energy comparisons

### LINKED ISSUE #886: GBSA test failures on FreeSolv dataset
**Describe the bug**
Some of the GBSA tests are reporting a different energy on a small (?) subset of the FreeSolv dataset. The differences look small to my eye but I'm not familiar enough with these tests to say they're not important.

These tests **only run** if OpenEye Toolkits are installed and "slow" tests are turned on via `--runslow`. These are automated on our nightly CI runs, but not the ones that run on each push and PR. These were previously missed because of #874 but not observed until last night's nighly because we do not run "slow" tests before merging. Presumably because of how long #874 lasted, there were several upstream updates, so maybe it's reasonable that some of these energies differ.

**To Reproduce**

Something like
```
pytest -v --runslow openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies
```
**Output**

Trimmed section of [log](https://pipelines.actions.githubusercontent.com/u2Wf9sTwFK9NZj8rRqcrkPO1UF3tPeGIdu4NysSHujaRyWbUAY/_apis/pipelines/1/runs/3036/signedlogcontent/5?urlExpires=2021-03-26T15%3A07%3A28.2666139Z&urlSigningMethod=HMACV1&urlSignature=NHb2pnozLBO4F6U5t0NdLryRLUjIAgtenMvLOdgPEeU%3D) 
```
2021-03-26T05:12:48.3733770Z =================================== FAILURES ===================================
2021-03-26T05:12:48.3736390Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[1328936-0_0_2-False-OBC1-True] _
2021-03-26T05:12:48.3738320Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3739680Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3741270Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3742500Z     atol=atol,
2021-03-26T05:12:48.3743720Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3745230Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3746450Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3747860Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3751510Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.09011694224912548 kJ/mol, potential energy 2: -0.09011567650344432 kJ/mol
2021-03-26T05:12:48.3754800Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[7859387-0_0_2-False-OBC1-False] _
2021-03-26T05:12:48.3756580Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3758230Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3769070Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3770170Z     atol=atol,
2021-03-26T05:12:48.3771360Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3772920Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3774260Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3775820Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3778030Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 0.41595119108834666 kJ/mol, potential energy 2: 0.4159649277135884 kJ/mol
2021-03-26T05:12:48.3782400Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[9434451-0_0_2-False-OBC1-True] _
2021-03-26T05:12:48.3784360Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3786100Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3787750Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3788780Z     atol=atol,
2021-03-26T05:12:48.3790000Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3791260Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3792460Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3793990Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3796700Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.04488653602139259 kJ/mol, potential energy 2: -0.04488564305646747 kJ/mol
2021-03-26T05:12:48.3799880Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[9534740-0_0_2-False-HCT-True] _
2021-03-26T05:12:48.3802190Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3803600Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3805420Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3806490Z     atol=atol,
2021-03-26T05:12:48.3807570Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3809250Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3810480Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3811730Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3814540Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.4004930038910004 kJ/mol, potential energy 2: -0.4005215052162612 kJ/mol
2021-03-26T05:12:48.3817770Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[9534740-0_0_4_fixed-False-HCT-True] _
2021-03-26T05:12:48.3819740Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3821280Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3822840Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3824010Z     atol=atol,
2021-03-26T05:12:48.3825080Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3826350Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3828060Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3829450Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3832540Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.4004930038910004 kJ/mol, potential energy 2: -0.40050457749714496 kJ/mol
2021-03-26T05:12:48.3836580Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[9741965-0_0_2-False-HCT-False] _
2021-03-26T05:12:48.3838370Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3840160Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3841450Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3842490Z     atol=atol,
2021-03-26T05:12:48.3843560Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3844830Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3846150Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3847510Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3850830Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.003823110028392307 kJ/mol, potential energy 2: -0.003820578243043826 kJ/mol
2021-03-26T05:12:48.3854250Z _ TestForceFieldParameterAssignment.test_freesolv_gbsa_energies[9741965-0_0_2-False-OBC1-False] _
2021-03-26T05:12:48.3856100Z openff/toolkit/tests/test_forcefield.py:3867: in test_freesolv_gbsa_energies
2021-03-26T05:12:48.3857480Z     off_omm_system, amber_omm_system, positions, by_force_type=False
2021-03-26T05:12:48.3858790Z openff/toolkit/tests/utils.py:677: in compare_system_energies
2021-03-26T05:12:48.3860190Z     atol=atol,
2021-03-26T05:12:48.3861340Z openff/toolkit/tests/utils.py:503: in compare_context_energies
2021-03-26T05:12:48.3862720Z     format_args=(potential_energy1, potential_energy2),
2021-03-26T05:12:48.3863920Z openff/toolkit/tests/utils.py:492: in raise_assert
2021-03-26T05:12:48.3865180Z     err_msg.format(*format_args), potential_energy1, potential_energy2
2021-03-26T05:12:48.3867960Z E   openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.25225082890618467 kJ/mol, potential energy 2: -0.2522448698537616 kJ/mol
2021-03-26T05:12:48.4048460Z =========================== short test summary info ============================
2021-03-26T05:12:48.4052280Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[1328936-0_0_2-False-OBC1-True]
2021-03-26T05:12:48.4055670Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[7859387-0_0_2-False-OBC1-False]
2021-03-26T05:12:48.4058910Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[9434451-0_0_2-False-OBC1-True]
2021-03-26T05:12:48.4062620Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[9534740-0_0_2-False-HCT-True]
2021-03-26T05:12:48.4065900Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[9534740-0_0_4_fixed-False-HCT-True]
2021-03-26T05:12:48.4069170Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[9741965-0_0_2-False-HCT-False]
2021-03-26T05:12:48.4072400Z FAILED openff/toolkit/tests/test_forcefield.py::TestForceFieldParameterAssignment::test_freesolv_gbsa_energies[9741965-0_0_2-False-OBC1-False]
2021-03-26T05:12:48.4074330Z = 7 failed, 25800 passed, 26 skipped, 81 xfailed, 315 xpassed, 4572 warnings in 12755.00s (3:32:34) =
```
**Computing environment (please complete the following information):**
 - Operating system
 - Output of running `conda list` 

**Additional context**
This is not due to (very) recent OpenEye release, version 2020.2 is still being pulled down as has been for a few weeks or months.

### PR DESCRIPTION
**Pull request by mattwthompson**
Fixes #886 

I'm tackling this since it's caused false-ish negatives in our CI for a few weeks.

I noticed that the energy differences are minuscule and tend to only show up when total energy of also small (i.e. -0.04488653602139259 vs -0.04488564305646747 kJ/mol). I checked the diff of serialized forces for one case and found that the values are identical to  around 10 significant digits:

```shell
<                 <Particle param1="-.09520000219345093" param2=".161" param3=".11592"/>
<                 <Particle param1="-.07689999788999557" param2=".161" param3=".11592"/>
<                 <Particle param1=".06949999928474426" param2=".161" param3=".11592"/>
<                 <Particle param1="-.2551000118255615" param2=".141" param3=".1128"/>
<                 <Particle param1=".03689999878406525" param2=".12100000000000001" param3=".10285000000000001"/>
<                 <Particle param1=".03689999878406525" param2=".12100000000000001" param3=".10285000000000001"/>
<                 <Particle param1=".03689999878406525" param2=".12100000000000001" param3=".10285000000000001"/>
<                 <Particle param1=".05220000073313713" param2=".12100000000000001" param3=".10285000000000001"/>
<                 <Particle param1=".05220000073313713" param2=".12100000000000001" param3=".10285000000000001"/>
<                 <Particle param1=".07129999995231628" param2=".12100000000000001" param3=".10285000000000001"/>
<                 <Particle param1=".07129999995231628" param2=".12100000000000001" param3=".10285000000000001"/>
---
>                 <Particle param1="-.09520000219511258" param2=".16099999999999998" param3=".11591999999999998"/>
>                 <Particle param1="-.07689999780488742" param2=".16099999999999998" param3=".11591999999999998"/>
>                 <Particle param1=".06949999945122184" param2=".16099999999999998" param3=".11591999999999998"/>
>                 <Particle param1="-.2551000120731192" param2=".141" param3=".1128"/>
>                 <Particle param1=".03689999879268808" param2=".12100000000000001" param3=".10285000000000001"/>
>                 <Particle param1=".03689999879268808" param2=".12100000000000001" param3=".10285000000000001"/>
>                 <Particle param1=".03689999879268808" param2=".12100000000000001" param3=".10285000000000001"/>
>                 <Particle param1=".05220000071341158" param2=".12100000000000001" param3=".10285000000000001"/>
>                 <Particle param1=".05220000071341158" param2=".12100000000000001" param3=".10285000000000001"/>
>                 <Particle param1=".0713" param2=".12100000000000001" param3=".10285000000000001"/>
>                 <Particle param1=".0713" param2=".12100000000000001" param3=".10285000000000001"/>
```

The actual comparison happens by calling [`np.allclose`](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html) with default arguments on computed energies, which attempts to account for both absolute (1e-8 tolerance) and relative (1e-5) errors.

> 
    If the following equation is element-wise True, then allclose returns True.

    absolute(a - b) <= (atol + rtol * absolute(b))

I don't think the default values are appropriate here. If `absolute(b)` is than 1e-2 or greater (as is the case in most systems, in kJ) then the relative tolerance term dominates and the criteria basically becomes "do my values differ within the first 5 significant digits?" But when energies approach 1e-3 it begins to compete with the absolute tolerance and the criteria becomes closer to "do my energies differ by more or less than 1e-8 kJ/mol? I'd argue that tolerance is overly strict for the MM corner of computational chemistry.

For example, assume noise on the order of 1e-6 - this doesn't show up in this formula when the energies are on the order of unity, but fails around 0.01:
```
>>> import numpy as np
>>> a, b = 1, 1 + 1e-6
>>> np.allclose(a, b)
True
>>> a, b = 0.01, 0.01 + 1e-6
>>> np.allclose(a, b)
False
>>> np.allclose(a, b, atol=1e-6)
True
```

Therefore, I think the test  failures are red herrings and to fix them I've loosened the absolute tolerance in the tests. I'm not sure what number to actually use; 1e-5 kJ/mol is a tiny energy for an MM force field but it might be too loose when comparing implementations that we think should be exact or close to exact. Maybe 1e-6 is a better choice?

I noticed that the tests failing on my machine aren't the same as the test that are failing on CI. Maybe this points to small platform differences in the OpenMM kernels? Again, I don't have a good sense of what sorts of numerical differences are typically reported.

- [x] Tag issue being addressed
- [ ] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/master/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable
- [ ] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

### CODE REVIEW

**Comment by mattwthompson** [comment]
One is still failing

```
<snip>
openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.4004836757641499 kJ/mol, potential energy 2: -0.40050407085760753 kJ/mol
 <snip>
openff.toolkit.tests.utils.FailedEnergyComparisonError: potential energy 1 -0.4004836757641499 kJ/mol, potential energy 2: -0.40050407085760753 kJ/mol
```

```
>>> import numpy as np
>>> a, b = -0.4004836757641499, -0.40050407085760753
>>> np.absolute(b - a)
2.0395093457636904e-05
>>> 1e-8 + np.absolute(b) * 1e-5
4.015040708576075e-06
```

Still an absolutely small energy difference but harder to sweep under the rug.

**Comment by j-wags** [comment]
Hm, so I actually get the same tests failing locally, but different values from both the previous ones, AND from the ones you posted above. My max deviation is 1.2e-4 kJ/mol.

One thing that might make this moot is the imminent implementation of partial charge rounding (#456). I'm testing that out to see whether some rounding magic will make this happy again, or blow this test entirely out of the water. In either case we may get to avoid facing this problem directly.

File: openff/toolkit/tests/test_forcefield.py [hunk 1]
```diff
@@ -3825,18 +3823,18 @@ def test_freesolv_gbsa_energies(
...
         assert abs(amber_energy[1] - off_energy[1]) < 1e-5 * unit.kilojoule / unit.mole
 
         # Ensure that all system energies are the same
         compare_system_energies(
-            off_omm_system, amber_omm_system, positions, by_force_type=False
+            off_omm_system,
+            amber_omm_system,
+            positions,
+            by_force_type=False,
+            atol=1.0e-5,
```
Comment (j-wags) [inline openff/toolkit/tests/test_forcefield.py:3837]:
```suggestion
            atol=1.0e-4,
```

**Comment by j-wags** [comment]
Ok, I'm just so gosh darn tired of seeing these tests fail on nightlies that I'm raising the tolerance. If tests pass now I'll revert the CI changes and merge.

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -145,9 +145,11 @@ jobs:
...
           PYTEST_ARGS+=" --ignore=openff/toolkit/tests/test_examples.py"
           PYTEST_ARGS+=" --ignore=openff/toolkit/tests/test_links.py"
-          if [[ "$GITHUB_EVENT_NAME" == "schedule" ]]; then
-            PYTEST_ARGS+=" --runslow"
-          fi
+          # REVERT BEFORE MERGE
+          # if [[ "$GITHUB_EVENT_NAME" == "schedule" ]]; then
+          #   PYTEST_ARGS+=" --runslow"
+          # fi
+          PYTEST_ARGS+=" --runslow"
```
Comment (j-wags) [inline .github/workflows/CI.yml:152]:
```suggestion
          if [[ "$GITHUB_EVENT_NAME" == "schedule" ]]; then
            PYTEST_ARGS+=" --runslow"
          fi
```

File: openff/toolkit/tests/test_forcefield.py [hunk 3]
```diff
@@ -3825,18 +3823,18 @@ def test_freesolv_gbsa_energies(
         amber_energy = get_context_potential_energy(amber_context, positions)
         off_energy = get_context_potential_energy(off_context, positions)
```
Comment (j-wags) [inline openff/toolkit/tests/test_forcefield.py:3825]: I am not a very clever person so I'd like to keep this in for when this problem comes back to our attention.

```suggestion
        # Very handy for debugging
        # print(openmm.XmlSerializer.serialize(off_gbsa_force))
        # print(openmm.XmlSerializer.serialize(amber_gbsa_force))
```

**Review by j-wags — APPROVED** [review]
with our apathies combined, I am captain planet. 

image


Well, I guess our CI will still burn the same amount of coal. But at least now the status will mean something.

**Pull request merged**